src/snovault/elasticsearch/indexer.py

```diff
@@ -143,140 +154,302 @@ def get_related_uuids(request, es, updated, renamed):
 def index(request):
-    INDEX = request.registry.settings['snovault.elasticsearch.index']
     # Setting request.datastore here only works because routed views are not traversed.
     request.datastore = 'database'
-    record = request.json.get('record', False)
-    dry_run = request.json.get('dry_run', False)
-    recovery = request.json.get('recovery', False)
-    es = request.registry[ELASTIC_SEARCH]
-    indexer = request.registry[INDEXER]
     session = request.registry[DBSESSION]()
     connection = session.connection()
-    first_txn = None
-    snapshot_id = None
-    restart=False
-    invalidated = []
-    xmin = -1
-    is_testing = asbool(request.registry.settings.get('testing', False))
-    is_testing_full = request.json.get('is_testing_full', False)
-
     # Currently 2 possible followup indexers (base.ini [set stage_for_followup = vis_indexer, region_indexer])
     stage_for_followup = list(request.registry.settings.get("stage_for_followup", '').replace(' ','').split(','))
-
     # May have undone uuids from prior cycle
-    state = IndexerState(es, INDEX, followups=stage_for_followup)
-
-    (xmin, invalidated, restart) = state.priority_cycle(request)
-    state.log_reindex_init_state()
-    # OPTIONAL: restart support
-    if restart:  # Currently not bothering with restart!!!
-        xmin = -1
-        invalidated = []
-    # OPTIONAL: restart support
-
-    result = state.get_initial_state()  # get after checking priority!
-
-    if xmin == -1 or len(invalidated) == 0:
-        xmin = get_current_xmin(request)
-
-        last_xmin = None
-        if 'last_xmin' in request.json:
-            last_xmin = request.json['last_xmin']
-        else:
-            status = es.get(index=INDEX, doc_type='meta', id='indexing', ignore=[400, 404])
-            if status['found'] and 'xmin' in status['_source']:
-                last_xmin = status['_source']['xmin']
-        if last_xmin is None:  # still!
-            if 'last_xmin' in result:
-                last_xmin = result['last_xmin']
-            elif 'xmin' in result and result['xmin'] < xmin:
-                last_xmin = result['state']
-
-        result.update(
-            xmin=xmin,
-            last_xmin=last_xmin,
-        )
-
-    if len(invalidated) > SEARCH_MAX:  # Priority cycle already set up
-        flush = True
-    else:
+    indexer_state = IndexerState(
+        request.registry[ELASTIC_SEARCH],
+        request.registry.settings['snovault.elasticsearch.index'],
+        followups=stage_for_followup,
+    )
+    # Check remote indexing and set head/indexing nodes
+    this_node = indexer_state.get_obj(INDEXING_NODE_INDEX)
+    other_node = indexer_state.get_obj(HEAD_NODE_INDEX)
+    remote_indexing = asbool(request.registry.settings.get('remote_indexing', False))
+    did_timeout = False
+    if remote_indexing:
+        this_node, other_node, continue_on, did_timeout = _get_nodes(request, indexer_state)
```

Comment:
yes, there is some duplication here and other places when dealing with the states.  Fiddling with it may cause bugs at this point.  It's pretty stable at the moment even with the deficiencies.  We discussed an overhaul of the indexer. Which is a better use of time than cleaning the thing I just tacked on here.